### PR TITLE
Fix non Aggregate compatibility

### PR DESCRIPTION
**Pull request by ggalmazor**
This PR solves issues when dealing with Ona servers:
- Can't connect to list forms in the pull tab
- Can't use the "start from last" feature when pulling forms
- Can't use the "pull before export" feature when exporting forms.

#### What has been done to verify that this works as intended?
I've tested this manually using my Ona account at https://odk.ona.io/ggalmazor
- Pull forms under various "start from last" scenarios: first time pull, second time pull, send new submissions and pull again, etc.
- Export forms with the "pull before export" option:
  - Single form
  - Multiple form, from Ona and Aggregate servers

#### Why is this the best possible solution? Were any other approaches considered?
Subtyping Cursor felt like the simplest way to go. The alternative would have been to revert the changes to Cursor that made it parse its value, but that would have an impact on how we deal with the "start from date" feature.

The changes to Cursor are required to get the submissions list when pulling forms. This is because Ona uses a simple numeric id, instead of a complex XML document like Aggregate.

An alternative to the changes to support the pull before export feature with Ona would be to simply not support it, but it felt like we ought to make the effort this time.  

#### How does this change affect users? Describe intentional changes to behavior and behavior that could have accidentally been affected by code changes. In other words, what are the regression risks?

Every time we change something related to the java prefs system we should be on the lookout for trouble.

This time, we need to save a new key with the cursor type (aggregate/ona) and when reading cursors from prefs, we base on that value to choose which factory to use (defaulting to aggregate)

If I did something wrong, users might have trouble with the "start from last" feature, and the "pull before export" feature, by extension.

#### Does this change require updates to documentation? If so, please file an issue at https://github.com/opendatakit/docs/issues/new and include the link below.
Nope.

### CODE REVIEW

**Comment by kkrawczyk123** [comment]
**Tested with success!**
 Verified on Mac, Ubuntu and Windows. Pulling from ona confirmed, pull before export and start pull from the last submission. Confirmed that no regression visible in pulling, pull before and start from last for Aggregate and in pulling from Central.  

@opendatakit-bot label "behavior verified"

File: src/org/opendatakit/briefcase/pull/aggregate/AggregateCursor.java [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+    );
+    return new AggregateCursor(cursorXml, Optional.of(lastUpdate), Optional.of(lastReturnedValue));
+  }
+
+  /**
+   * Returns a synthetic Cursor instance with the provided values.
+   */
+  public static AggregateCursor of(LocalDate date) {
+    OffsetDateTime lastUpdate = date.atStartOfDay().atOffset(ZoneOffset.UTC);
+    String cursorXml = String.format("<cursor xmlns=\"http://www.opendatakit.org/cursor\">" +
```
Comment (dcbriccetti) [inline src/org/opendatakit/briefcase/pull/aggregate/AggregateCursor.java]: Could we remove this duplication, please?
Screen Shot 2019-06-30 at 11 02 01 AM
```
  /**
   * Returns a synthetic Cursor instance with the provided values.
   */
  public static AggregateCursor of(OffsetDateTime lastUpdate, String lastReturnedValue) {
    String cursorXml = getCursorXml(lastUpdate, lastReturnedValue);
    return new AggregateCursor(cursorXml, Optional.of(lastUpdate), Optional.of(lastReturnedValue));
  }

  /**
   * Returns a synthetic Cursor instance with the provided values.
   */
  public static AggregateCursor of(LocalDate date) {
    OffsetDateTime lastUpdate = date.atStartOfDay().atOffset(ZoneOffset.UTC);
    String cursorXml = getCursorXml(lastUpdate, "");
    return new AggregateCursor(cursorXml, Optional.of(lastUpdate), Optional.empty());
  }

  private static String getCursorXml(OffsetDateTime lastUpdate, String lastReturnedValue) {
    return String.format("<cursor xmlns=\"http://www.opendatakit.org/cursor\">" +
            "<attributeName>_LAST_UPDATE_DATE</attributeName>" +
            "<attributeValue>%s</attributeValue>" +
            "<uriLastReturnedValue>%s</uriLastReturnedValue>" +
            "<isForwardCursor>true</isForwardCursor>" +
            "</cursor>",
        lastUpdate.format(ISO_OFFSET_DATE_TIME),
        lastReturnedValue);
  }
```

File: src/org/opendatakit/briefcase/reused/http/CommonsHttp.java [hunk 2]
```diff
@@ -78,7 +78,7 @@ private static HttpClientBuilder getBaseBuilder(int maxConnections) {
   @Override
   public <T> Response<T> execute(Request<T> request) {
     // Ensure that we will be using fresh credentials
-    executor.clearAuth();
+    //executor.clearAuth();
```
Comment (dcbriccetti) [inline src/org/opendatakit/briefcase/reused/http/CommonsHttp.java]: Remove?

File: src/org/opendatakit/briefcase/reused/http/CommonsHttp.java [hunk 3]
```diff
@@ -67,7 +67,7 @@ private static HttpClientBuilder getBaseBuilder(int maxConnections) {
         .create()
         .setMaxConnPerRoute(maxConnections)
         .setMaxConnTotal(maxConnections)
-        .disableAuthCaching()
+        //.disableAuthCaching()
```
Comment (dcbriccetti) [inline src/org/opendatakit/briefcase/reused/http/CommonsHttp.java]: Remove?

File: src/org/opendatakit/briefcase/reused/http/RequestBuilder.java [hunk 4]
```diff
@@ -220,9 +220,14 @@ public static boolean isUri(String candidate) {
...
-      throw new BriefcaseException("Can't apply withPath() twice");
-    URL newBaseUrl = url(baseUrl + (path.startsWith("/") ? path : "/" + path));
+    String cleanPath = path.startsWith("/") && path.endsWith("/")
+        ? path.substring(1, path.length() - 1)
+        : path.startsWith("/")
+        ? path.substring(1)
+        : path.endsWith("/")
+        ? path.substring(0, path.length() - 1)
+        : path;
+    URL newBaseUrl = url(baseUrl + "/" + cleanPath);
```
Comment (dcbriccetti) [inline src/org/opendatakit/briefcase/reused/http/RequestBuilder.java]: Is this more readable?
Screen Shot 2019-06-30 at 11 44 38 AM

```
    int start = path.startsWith("/") ? 1 : 0;
    int end   = path.endsWith  ("/") ? 1 : 0;
    String cleanPath = path.substring(start, path.length() - end);
```

Comment (dcbriccetti) [reply src/org/opendatakit/briefcase/reused/http/RequestBuilder.java]: startOffset and endOffset would be better

File: src/org/opendatakit/briefcase/pull/aggregate/AggregateCursor.java (hunk 1 above)
Comment (ggalmazor) [reply src/org/opendatakit/briefcase/pull/aggregate/AggregateCursor.java]: We can't remove duplication exactly as you propose, but we can make it better. It's key we have a `<tag/>`, as opposed to the `<tag></tag>` you're proposing. Don't ask me why but I've noticed different behavior with `<tag/>`.

I'll work in the extraction as you've laid it out, but with that detail on mind.

File: src/org/opendatakit/briefcase/reused/http/RequestBuilder.java (hunk 4 above)
Comment (ggalmazor) [reply src/org/opendatakit/briefcase/reused/http/RequestBuilder.java]: Much better, thanks!

**Comment by ggalmazor** [comment]
Thanks, @dcbriccetti! All your comments are super helpful. I've implemented the three changes you have suggested. Could you do a sanity check on them, please?

File: src/org/opendatakit/briefcase/pull/aggregate/AggregateCursor.java [hunk 5]
```diff
@@ -128,6 +113,20 @@ public static AggregateCursor of(LocalDate date, String lastReturnedValue) {
...
+    return String.format("<cursor xmlns=\"http://www.opendatakit.org/cursor\">" +
+            "<attributeName>_LAST_UPDATE_DATE</attributeName>" +
+            "<attributeValue>%s</attributeValue>" +
+            "%s" +
+            "<isForwardCursor>true</isForwardCursor>" +
+            "</cursor>",
+        lastUpdate.format(ISO_OFFSET_DATE_TIME),
+        lastUri
+            .map(value -> "<uriLastReturnedValue>" + value + "</uriLastReturnedValue>")
+            .orElse("<uriLastReturnedValue/>")
```
Comment (dcbriccetti) [inline src/org/opendatakit/briefcase/pull/aggregate/AggregateCursor.java]: Comment that the `<uriLastReturnedValue/>` must be in that form, to help future readers?

**Comment by dcbriccetti** [comment]
Looks good!

**Pull request merged**